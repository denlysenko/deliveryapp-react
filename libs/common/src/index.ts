export * from './lib/constants';
export * from './lib/errors';
export * from './lib/roles';
export * from './lib/app-menu';
export * from './lib/messages';
export * from './lib/payment-methods';
