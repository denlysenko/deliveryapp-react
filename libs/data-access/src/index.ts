import * as authClient from './lib/api/auth';
import * as usersClient from './lib/api/users';
import * as ordersClient from './lib/api/orders';
import * as paymentsClient from './lib/api/payments';

export { authClient, usersClient, ordersClient, paymentsClient };

export * from './lib/models/auth';
export * from './lib/models/user';
export * from './lib/models/validation-error';
export * from './lib/models/order';
export * from './lib/models/list-response';
export * from './lib/models/orders-filter';
export * from './lib/models/payment';
export * from './lib/models/payments-filter';
export * from './lib/models/users-filter';

export * from './lib/context/auth';
export * from './lib/context/orders';
export * from './lib/context/payments';
