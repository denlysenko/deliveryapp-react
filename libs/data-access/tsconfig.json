{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "jsx": "react",
    "esModuleInterop": true,
    "types": ["node", "jest"]
  },
  "include": ["**/*.ts"]
}
