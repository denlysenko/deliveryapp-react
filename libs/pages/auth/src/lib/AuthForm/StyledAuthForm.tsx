import styled from 'styled-components';

export const StyledAuthForm = styled.div`
  .p-col-12 {
    padding: 15px 100px;

    @media (max-width: 640px) {
      padding: 15px 50px;
    }
  }

  .error-message {
    padding: 0 100px;
  }

  .input-wrapper {
    position: relative;

    input {
      width: 100%;
    }

    .fa {
      position: absolute;
      width: 14px;
      right: 8px;
      font-size: 14px;
      color: #dee4e9;
    }

    .fa-user-circle-o,
    .fa-building-o,
    .fa-phone {
      top: 7px;
    }

    .fa-lock,
    .fa-envelope-o {
      top: 6px;
    }
  }

  button {
    padding: 0;
    width: 100%;
    display: block;
    margin-bottom: 20px;
  }
`;
