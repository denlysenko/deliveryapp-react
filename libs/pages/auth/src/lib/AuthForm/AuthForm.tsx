import React, { useEffect, useRef } from 'react';
import { useFormik } from 'formik';

import { AuthCredentials, ValidationError } from '@deliveryapp/data-access';

import { Button } from 'primereact/button';
import { InputMask } from 'primereact/inputmask';
import { InputText } from 'primereact/inputtext';
import { Messages } from 'primereact/messages';
import { Message } from 'primereact/message';

import * as Yup from 'yup';
import { isNil } from 'lodash-es';

import { StyledAuthForm } from './StyledAuthForm';

interface LoginError {
  message: string;
}

export interface AuthFormProps {
  isLoggingIn: boolean;
  loading: boolean;
  error: LoginError | ValidationError | null;
  onFormSubmit: (body: Partial<AuthCredentials>) => void;
}

const ValidationSchema = Yup.object().shape({
  email: Yup.string()
    .email('Invalid email')
    .required('Required field'),
  password: Yup.string().required('Required field')
});

export const AuthForm = (props: AuthFormProps) => {
  const { isLoggingIn, loading, error, onFormSubmit } = props;
  const messages = useRef(null);

  const formik = useFormik({
    initialValues: {
      firstName: '',
      lastName: '',
      email: '',
      company: '',
      phone: '',
      password: ''
    },
    validationSchema: ValidationSchema,
    onSubmit: values => {
      if (isLoggingIn) {
        const { email, password } = values;
        onFormSubmit({ email, password });
      } else {
        onFormSubmit(values);
      }
    }
  });

  useEffect(() => {
    formik.setErrors({});
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isLoggingIn]);

  useEffect(() => {
    if (!isNil(error)) {
      if ('errors' in error) {
        error.errors.forEach(err => {
          formik.setFieldError(err.path, err.message);
        });
      } else {
        messages.current.show({
          severity: 'error',
          summary: error.message,
          closable: false
        });
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [error]);

  return (
    <StyledAuthForm>
      <form onSubmit={formik.handleSubmit} autoComplete="off" noValidate>
        <Messages className="error-message" ref={messages} />
        {!isLoggingIn && (
          <>
            <div className="ui-g-12">
              <div className="input-wrapper p-float-label">
                <InputText
                  id="firstName"
                  value={formik.values.firstName}
                  onChange={formik.handleChange}
                />
                <label htmlFor="firstName">First Name</label>
                <i className="fa fa-user-circle-o"></i>
              </div>
            </div>
            <div className="ui-g-12">
              <div className="input-wrapper p-float-label">
                <InputText
                  id="lastName"
                  value={formik.values.lastName}
                  onChange={formik.handleChange}
                />
                <label htmlFor="lastName">Last Name</label>
                <i className="fa fa-user-circle-o"></i>
              </div>
            </div>
            <div className="ui-g-12">
              <div className="input-wrapper p-float-label">
                <InputText
                  id="company"
                  value={formik.values.company}
                  onChange={formik.handleChange}
                />
                <label htmlFor="company">Company</label>
                <i className="fa fa-building-o"></i>
              </div>
            </div>
            <div className="ui-g-12">
              <div className="input-wrapper p-float-label">
                <InputMask
                  id="phone"
                  mask="(999) 999-9999"
                  value={formik.values.phone}
                  onChange={e => formik.setFieldValue('phone', e.value)}
                />
                <label htmlFor="phone">Phone</label>
                <i className="fa fa-phone"></i>
              </div>
            </div>
          </>
        )}
        <div className="ui-g-12">
          <div className="input-wrapper p-float-label">
            <InputText
              id="email"
              className={
                formik.touched.email && formik.errors.email ? 'invalid' : ''
              }
              value={formik.values.email}
              onChange={formik.handleChange}
            />
            <label htmlFor="email">Email</label>
            <i className="fa fa-user-circle-o"></i>
          </div>
          {formik.touched.email && formik.errors.email && (
            <Message severity="error" text={formik.errors.email}></Message>
          )}
        </div>
        <div className="ui-g-12">
          <div className="input-wrapper p-float-label">
            <InputText
              type="password"
              id="password"
              className={
                formik.touched.password && formik.errors.password
                  ? 'invalid'
                  : ''
              }
              value={formik.values.password}
              onChange={formik.handleChange}
            />
            <label htmlFor="password">Password</label>
            <i className="fa fa-lock"></i>
          </div>
          {formik.touched.password && formik.errors.password && (
            <Message severity="error" text={formik.errors.password}></Message>
          )}
        </div>
        <div className="ui-g-12 button-container">
          <Button
            type="submit"
            label={isLoggingIn ? 'Login' : 'Register'}
            className="p-button-raised"
            disabled={loading}
          />
        </div>
      </form>
    </StyledAuthForm>
  );
};
