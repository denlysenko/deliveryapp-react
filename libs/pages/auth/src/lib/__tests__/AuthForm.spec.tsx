/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { render, fireEvent, cleanup, waitFor } from '@testing-library/react';

import { ERRORS } from '@deliveryapp/common';
import { ValidationError } from '@deliveryapp/data-access';

import { AuthForm } from '../AuthForm/AuthForm';

const email = 'test@test.com';
const password = 'password';

const props = {
  isLoggingIn: true,
  loading: false,
  error: null,
  onFormSubmit: jest.fn()
};

describe('AuthForm', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  describe('Login', () => {
    it('should render login form', () => {
      const { baseElement } = render(<AuthForm {...props} />);
      expect(baseElement).toMatchSnapshot();
    });

    it('should have Login label on submit button', () => {
      const { getByTestId } = render(<AuthForm {...props} />);
      expect(getByTestId('submit')).toHaveTextContent('Login');
    });

    it('should show error message when error prop passed', async () => {
      const { container, rerender, getByTestId } = render(
        <AuthForm {...props} />
      );

      fireEvent.change(getByTestId('email'), {
        target: {
          value: email
        }
      });

      fireEvent.change(getByTestId('password'), {
        target: {
          value: password
        }
      });

      fireEvent.click(getByTestId('submit'));

      await waitFor(() => {});

      const errorMessage = 'Error message';

      rerender(<AuthForm {...props} error={{ message: errorMessage }} />);

      expect(container.querySelector('#error-message')).toHaveTextContent(
        errorMessage
      );
    });
  });

  describe('Register', () => {
    it('should render register form', () => {
      const { baseElement } = render(
        <AuthForm {...props} isLoggingIn={false} />
      );
      expect(baseElement).toMatchSnapshot();
    });

    it('should have Register label on submit button', () => {
      const { getByTestId } = render(
        <AuthForm {...props} isLoggingIn={false} />
      );
      expect(getByTestId('submit')).toHaveTextContent('Register');
    });

    it('should show error messages when error prop passed', async () => {
      const emailErrorMessage = 'Email error';
      const passwordErrorMessage = 'Password error';

      const error = {
        errors: [
          { path: 'email', message: emailErrorMessage },
          { path: 'password', message: passwordErrorMessage }
        ]
      } as ValidationError;

      const { container, getByTestId } = render(
        <AuthForm {...props} isLoggingIn={false} error={error} />
      );

      expect(container.querySelector('#email-error')).toHaveTextContent(
        emailErrorMessage
      );

      expect(container.querySelector('#password-error')).toHaveTextContent(
        passwordErrorMessage
      );

      fireEvent.change(getByTestId('email'), {
        target: {
          value: email
        }
      });

      fireEvent.change(getByTestId('password'), {
        target: {
          value: password
        }
      });

      await waitFor(() => {
        expect(container.querySelector('#email-error')).not.toBeInTheDocument();

        expect(
          container.querySelector('#password-error')
        ).not.toBeInTheDocument();
      });
    });
  });

  describe('Validations', () => {
    describe('email', () => {
      it('should display required error', async () => {
        const { getByTestId, container } = render(<AuthForm {...props} />);

        fireEvent.click(getByTestId('submit'));

        await waitFor(() => {
          expect(container.querySelector('#email-error')).toHaveTextContent(
            ERRORS.REQUIRED_FIELD
          );
        });
      });

      it('should display invalid email error', async () => {
        const { getByTestId, container } = render(<AuthForm {...props} />);

        fireEvent.change(getByTestId('email'), {
          target: {
            value: 'test'
          }
        });

        fireEvent.click(getByTestId('submit'));

        await waitFor(() => {
          expect(container.querySelector('#email-error')).toHaveTextContent(
            ERRORS.INVALID_EMAIL
          );
        });
      });

      it('should not display any errors', async () => {
        const { getByTestId, container } = render(<AuthForm {...props} />);

        fireEvent.change(getByTestId('email'), {
          target: {
            value: email
          }
        });

        fireEvent.click(getByTestId('submit'));

        await waitFor(() => {
          expect(
            container.querySelector('#email-error')
          ).not.toBeInTheDocument();
        });
      });
    });

    describe('password', () => {
      it('should display required error', async () => {
        const { getByTestId, container } = render(<AuthForm {...props} />);

        fireEvent.click(getByTestId('submit'));

        await waitFor(() => {
          expect(container.querySelector('#password-error')).toHaveTextContent(
            ERRORS.REQUIRED_FIELD
          );
        });
      });

      it('should not display any errors', async () => {
        const { getByTestId, container } = render(<AuthForm {...props} />);

        fireEvent.change(getByTestId('password'), {
          target: {
            value: password
          }
        });

        fireEvent.click(getByTestId('submit'));

        await waitFor(() => {
          expect(
            container.querySelector('#password-error')
          ).not.toBeInTheDocument();
        });
      });
    });

    it('should reset errors when isLogginIn prop has changed', async () => {
      const { getByTestId, container, rerender } = render(
        <AuthForm {...props} />
      );

      fireEvent.click(getByTestId('submit'));

      await waitFor(() => {
        expect(container.querySelector('#email-error')).toHaveTextContent(
          ERRORS.REQUIRED_FIELD
        );
      });

      rerender(<AuthForm {...props} isLoggingIn={false} />);

      expect(container.querySelector('#email-error')).not.toBeInTheDocument();
    });
  });

  describe('Submitting', () => {
    it('should call onFormSubmit prop', async () => {
      const { getByTestId } = render(<AuthForm {...props} />);

      fireEvent.change(getByTestId('email'), {
        target: {
          value: email
        }
      });

      fireEvent.change(getByTestId('password'), {
        target: {
          value: password
        }
      });

      fireEvent.click(getByTestId('submit'));

      await waitFor(() => {
        expect(props.onFormSubmit).toBeCalledTimes(1);
        expect(props.onFormSubmit).toBeCalledWith({
          email,
          password,
          company: '',
          firstName: '',
          lastName: '',
          phone: ''
        });
      });
    });

    it('should disable button when loading prop passed', async () => {
      const { getByTestId } = render(<AuthForm {...props} loading={true} />);
      expect(getByTestId('submit')).toBeDisabled();
    });
  });
});
