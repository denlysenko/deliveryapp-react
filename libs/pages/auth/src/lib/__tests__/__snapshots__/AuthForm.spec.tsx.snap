// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm Login should render login form 1`] = `
<body>
  <div>
    <div
      class="sc-AxjAm kyfgSK"
    >
      <form
        autocomplete="off"
        novalidate=""
      >
        <div
          class="error-message"
          id="error-message"
        >
          <div />
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="email"
              id="email"
              name="email"
              value=""
            />
            <label
              for="email"
            >
              Email
            </label>
            <i
              class="fa fa-user-circle-o"
            />
          </div>
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <label
              for="password"
            >
              Password
            </label>
            <i
              class="fa fa-lock"
            />
          </div>
        </div>
        <div
          class="p-col-12 button-container"
        >
          <button
            class="p-button p-component p-button-raised p-button-text-only"
            data-testid="submit"
            type="submit"
          >
            <span
              class="p-button-text p-c"
            >
              Login
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;

exports[`AuthForm Register should render register form 1`] = `
<body>
  <div>
    <div
      class="sc-AxjAm kyfgSK"
    >
      <form
        autocomplete="off"
        novalidate=""
      >
        <div
          class="error-message"
          id="error-message"
        >
          <div />
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="firstName"
              id="firstName"
              name="firstName"
              value=""
            />
            <label
              for="firstName"
            >
              First Name
            </label>
            <i
              class="fa fa-user-circle-o"
            />
          </div>
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="lastName"
              id="lastName"
              name="lastName"
              value=""
            />
            <label
              for="lastName"
            >
              Last Name
            </label>
            <i
              class="fa fa-user-circle-o"
            />
          </div>
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="company"
              id="company"
              name="company"
              value=""
            />
            <label
              for="company"
            >
              Company
            </label>
            <i
              class="fa fa-building-o"
            />
          </div>
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              id="phone"
              name="phone"
              type="text"
            />
            <label
              for="phone"
            >
              Phone
            </label>
            <i
              class="fa fa-phone"
            />
          </div>
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="email"
              id="email"
              name="email"
              value=""
            />
            <label
              for="email"
            >
              Email
            </label>
            <i
              class="fa fa-user-circle-o"
            />
          </div>
        </div>
        <div
          class="p-col-12"
        >
          <div
            class="input-wrapper p-float-label"
          >
            <input
              class="p-inputtext p-component"
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <label
              for="password"
            >
              Password
            </label>
            <i
              class="fa fa-lock"
            />
          </div>
        </div>
        <div
          class="p-col-12 button-container"
        >
          <button
            class="p-button p-component p-button-raised p-button-text-only"
            data-testid="submit"
            type="submit"
          >
            <span
              class="p-button-text p-c"
            >
              Register
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
